browser_patches/firefox/juggler/content/Runtime.js

```diff
@@ -505,9 +493,71 @@ class ExecutionContext {
   }
 
   _serialize(obj) {
-    const result = this._debuggee.executeInGlobalWithBindings('stringify(e)', {e: obj, stringify: this._jsonStringifyObject});
-    if (result.throw)
+    function serialize(root, jsonStringify, isFinite, isArray, getObjectEntries) {
+      function isInvalidValue(value) {
+        const type = typeof value;
+        return type === 'symbol' || type === 'function' || type === 'undefined';
+      }
+
+      function serializeArray(array) {
+        let result = '';
+        for (let i = 0; i < array.length; ++i) {
+          if (result)
+            result += ',';
+          result += isInvalidValue(array[i]) ? 'null' : innerSerialize(array[i]);
+        }
+        return '[' + result + ']'
+      }
+
+      function serializeObject(obj) {
+        let result = '';
+        for (const [key, value] of getObjectEntries(obj)) {
+          if (isInvalidValue(value))
+            continue;
+          if (result)
+            result += ',';
+          result += jsonStringify(key) + ':' + innerSerialize(value);
+        }
+        return '{' + result + '}';
+      }
+
+      const visitedObjects = new Set();
+      function innerSerialize(value) {
+        const type = typeof value;
+        if (type === 'string')
+          return jsonStringify(value);
+        if (type === 'number')
+          return isFinite(value) ? value + '' : 'null';
+        if (type === 'boolean')
+          return value + '';
+        // null type is 'object' - handle proactively.
+        if (value === null)
+          return 'null';
+        if (type === 'object') {
+          if (visitedObjects.has(value))
+            throw new Error('Cannot serialize circular structure');
+          visitedObjects.add(value);
+          return isArray(value) ? serializeArray(value) : serializeObject(value);
+        }
+        throw new Error('ERROR: do not know how to serialize ' + type);
+      }
+
+      if (isInvalidValue(root))
+        return undefined;
+      return innerSerialize(root);
+    }
+
+    const result = this._debuggee.executeInGlobalWithBindings(`(${serialize.toString()})(root, jsonStringify, isFinite, isArray, getObjectEntries)`, {
```

Comment:
let's eval serialize in the constructor too similar to the other bindings?

Reply:
done